osu.Game/Screens/Menu/MenuVisualisation.cs

```diff
@@ -1,11 +1,128 @@
 using osu.Framework.Graphics;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Configuration;
+using osu.Framework.Graphics.Containers;
+using OpenTK;
+using OpenTK.Graphics;
+using System;
 
 namespace osu.Game.Screens.Menu
 {
-    internal class MenuVisualisation : Drawable
+    internal class MenuVisualisation : Container
     {
+        public override bool HandleInput => false;
+
+        private Bindable<WorkingBeatmap> beatmap;
+        private Bindable<bool> kiai = new Bindable<bool>();
+
+        private int indexOffset = 0;
+        private double timeOfLastUpdate = double.MinValue;
+        private float[] audioData;
+
+        private const int bars_per_visualizer = 250;
+
+        public MenuVisualisation()
+        {
+            audioData = new float[256];
+            Size = new Vector2(460);
+            Anchor = Anchor.Centre;
+            Origin = Anchor.Centre;
+
+            float barWidth = Size.X * (float)Math.Sqrt(2f * (1f - Math.Cos(MathHelper.DegreesToRadians(360f / bars_per_visualizer)))) / 2.2f;
+
+            for (int i = 0; i < 5; i++)
+                for (int j = 0; j < bars_per_visualizer; j++)
+                {
+                    Add(new Box()
+                    {
+                        Anchor = Anchor.BottomCentre,
+                        Origin = Anchor.BottomCentre,
+                        Size = new Vector2(barWidth, 300),
+                        RelativePositionAxes = Axes.Both,
+                        Scale = new Vector2(1, 0),
+                        Colour = Color4.White,
+                        Alpha = 0.2f,
+                        Position = new Vector2(
+                        -(float)Math.Sin((float)(j + i * 300) / bars_per_visualizer * 2 * MathHelper.Pi) / 2,
+                        -0.5f + (float)Math.Cos((float)(j + i * 300) / bars_per_visualizer * 2 * MathHelper.Pi) / 2),
+                        Rotation = 360f / bars_per_visualizer * j + 180 + 360f / 5 * i,
+                    });
+                    //Add(new QuintVisualisationBar(barWidth) { Rotation = (360f / bars_around_circle * i) });
+                }
+
+            kiai.ValueChanged += updateKiai;
+        }
+
+        private void updateKiai(bool newValue)
+        {
+            FadeTo(newValue ? 2.5f : 1, 100);
+        }
+
+        protected override void Update()
+        {
+            if (beatmap?.Value != null)
+            {
+                if ((bool)beatmap.Value?.Track?.IsRunning)
+                {
+                    ControlPoint kiaiControlPoint;
+                    beatmap.Value.Beatmap.TimingInfo.TimingPointAt(beatmap.Value.Track.CurrentTime, out kiaiControlPoint);
+                    kiai.Value = (kiaiControlPoint?.KiaiMode ?? false);
+
+
+                    if (timeOfLastUpdate + 50 <= Clock.CurrentTime)
+                    {
+                        int i = 0;
+
+                        audioData = (beatmap.Value.Track as TrackBass).GetChannelData();

```

Comment:
Raw bass methods should never be called from the main thread. Probably should make `GetChannelData` an internal thing to the framework, store its result in a local variable, and copy that local variable whenever it's requested from here. Similar to how other bass calls are used to update values inside the `Update` method of the audio tracks (which is always invoked from the audio thread).

That aside, your code here will crash if `Track` is not `TrackBass` which is not okay. It should be implementation agnostic.